File: code/IC10/backend/compiler/tests/node/__tests__/system.test.ts
/*
 * Copyright (c) 2026. All rights reserved.
 * This source code is licensed under the CC BY-NC-SA
 * (Creative Commons Attribution-NonCommercial-NoDerivatives) License, By Xiao Songtao.
 * This software is protected by copyright law. Reproduction, distribution, or use for commercial
 * purposes is prohibited without the author's permission. If you have any questions or require
 * permission, please contact the author: 2207150234@st.sziit.edu.cn
 */
/**
 * @file system.test.ts
 * @brief IC10 Node.js 绑定 - 系统测试
 * @details 端到端测试完整 IC10 程序的编译流程，包括：
 *          - 真实 IC10 程序（气闸控制、气体过滤、计数循环）
 *          - 完整指令集覆盖
 *          - 预处理指令组合
 *          - 错误场景（未定义变量、重定义）
 *          - 边界情况（空程序、纯注释、超大程序）
 *          - 前向引用解析
 *          - HASH/STR 宏
 */
import IC10NodeAPI = require('ic10c-node');
import {setupUTF8Console} from '../utils';
import * as fs from 'fs';
import * as path from 'path';

const {IC10CompilerLocal, TokenType, Lexer, Program, Parser, Analyser, SymbolTable, BasicType} = IC10NodeAPI;

beforeAll(() => {
    setupUTF8Console();
    if (typeof IC10CompilerLocal.setLanguage === 'function')
        IC10CompilerLocal.setLanguage('zh-hans');
});

// ============================================================
// 辅助函数和类型
// ============================================================

interface CompileResult {
    tokens: ReturnType<typeof Lexer.tokenize>;
    program: InstanceType<typeof Program>;
    parser: InstanceType<typeof Parser>;
    analyser: InstanceType<typeof Analyser>;
}

/** 完整编译流程：源码 → Token → AST → 语义分析 */
async function compile(source: string): Promise<CompileResult> {
    const tokens = Lexer.tokenize(source);
    const parser = new Parser(tokens);
    const program = parser.parse();
    const analyser = new Analyser();
    await analyser.visit(program);
    return {tokens, program, parser, analyser};
}

/** 读取 grammarTest.ic 文件（基于 __dirname 解析路径） */
function readGrammarTestFile(): string {
    // __tests__ → ..(node) → ..(tests) → ..(compiler) → ..(backend) → ..(IC10) → assets → ic → grammarTest.ic
    const filePath = path.join(__dirname, '..', '..', '..', '..', '..', 'assets', 'ic', 'grammarTest.ic');
    return fs.readFileSync(filePath, 'utf-8');
}

// ============================================================
// 真实 IC10 程序测试
// ============================================================

describe('Real IC10 programs', () => {
    it('should compile counting loop program', async () => {
        const source = [
            'alias counter r0',
            'define LIMIT 10',
            'start:',
            'move r0 0',
            'loop:',
            'add r0 r0 1',
            'blt r0 LIMIT loop',
            'yield',
            'j start',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
        expect(result.program.statements).toHaveLength(9);
    });

        });

// ============================================================
// grammarTest.ic 文件测试
// ============================================================

describe('grammarTest.ic fixture', () => {
    it('should compile grammarTest.ic successfully', async () => {
        const source = readGrammarTestFile();
        expect(source).toBeTruthy();

        const result = await compile(source);

        expect(result.program.statements.length).toBeGreaterThan(0);
    });

        });

// ============================================================
// 指令元数完整覆盖测试
// ============================================================

describe('All instruction arities', () => {
    it('should compile program with all instruction arities', async () => {
        // NOTE: lbn/lbns的最后操作数需为identifier|number(logicType/batchMode)，不能用register
        const source = [
            '# 零元',
            'hcf',
            'yield',
            '# 一元',
            'sleep 100',
            'j main',
            '# 二元',
            'move r0 0',
            'abs r0 r1',
            '# 三元',
            'add r0 r1 r2',
            'sub r0 r1 r2',
            'mul r0 r1 r2',
            '# 四元',
            'lerp r0 r1 r2 r3',
            '# 五元',
            'lbn r0 100 200 2 Average',
            '# 六元',
            'lbns r0 100 200 3 Quantity Average',
            'main:',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
        expect(result.program.statements.length).toBeGreaterThanOrEqual(12);
    });
});

// ============================================================
// 预处理指令完整覆盖测试
// ============================================================

describe('All preprocessor directives', () => {
    it('should compile program with all preprocessor directive types', async () => {
        const source = [
            'alias devA d0',
            'alias regA r0',
            'define COUNT 42',
            'define HEX_VAL $FF',
            'define HASH_VAL HASH("StructureLiquidVolumePump")',
            'define STR_VAL STR("SomeHash")',
            'main:',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// 错误场景测试
// ============================================================

describe('Error scenarios', () => {
                    });

// ============================================================
// 边界情况测试
// ============================================================

describe('Boundary cases', () => {
    it('should handle empty source', async () => {
        const result = await compile('');

        expect(result.program.statements).toHaveLength(0);
        expect(result.parser.diagnostics).toHaveLength(0);
    });

    it('should handle source with only comments', async () => {
        const source = '# comment 1\n# comment 2\n# comment 3\n';
        const result = await compile(source);

        expect(result.program.statements).toHaveLength(0);
        expect(result.parser.diagnostics).toHaveLength(0);
    });

    it('should handle source with only newlines', async () => {
        const source = '\n\n\n\n';
        const result = await compile(source);

        expect(result.program.statements).toHaveLength(0);
    });

    it('should handle source with trailing newlines', async () => {
        const source = 'hcf\n\n\n\n';
        const result = await compile(source);

        expect(result.program.statements).toHaveLength(1);
        expect(result.parser.diagnostics).toHaveLength(0);
    });

    it('should handle single statement without trailing newline', async () => {
        const source = 'hcf';
        const result = await compile(source);

        expect(result.program.statements).toHaveLength(1);
        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// 前向引用测试
// ============================================================

describe('Forward references', () => {
    it('should resolve forward label reference in jump instruction', async () => {
        const source = [
            'j target',
            'move r0 1',
            'target:',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        // 前向引用应被解析，不应有未定义标识符错误
        const undefinedErrors = result.analyser.diagnostics.filter(
            (d: any) => d.id.includes('IEA3')
        );
        expect(undefinedErrors).toHaveLength(0);
    });

        it('should resolve multiple forward references', async () => {
        const source = [
            'j label1',
            'j label2',
            'j label3',
            'label1:',
            'hcf',
            'label2:',
            'yield',
            'label3:',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        const undefinedErrors = result.analyser.diagnostics.filter(
            (d: any) => d.id.includes('IEA3')
        );
        expect(undefinedErrors).toHaveLength(0);
    });
});

// ============================================================
// HASH 和 STR 宏测试
// ============================================================

describe('HASH and STR macros', () => {
    it('should compile HASH macro in define', async () => {
        const source = 'define PUMP_HASH HASH("StructureLiquidVolumePump")\nhcf\n';
        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });

    it('should compile STR macro in define', async () => {
        const source = 'define LABEL STR("SomeHash")\nhcf\n';
        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });

    it('should compile multiple HASH and STR macros', async () => {
        const source = [
            'define HASH1 HASH("StructureLiquidVolumePump")',
            'define HASH2 HASH("StructureGasVentSensor")',
            'define STR1 STR("Label1")',
            'define STR2 STR("Label2")',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// 设备引用测试
// ============================================================

describe('Device references', () => {
    it('should compile program using all device references d0-d5', async () => {
        const source = [
            'l r0 d0 Setting',
            'l r1 d1 Setting',
            'l r2 d2 Setting',
            'l r3 d3 Setting',
            'l r4 d4 Setting',
            'l r5 d5 Setting',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });

    });

// ============================================================
// 寄存器边界测试
// ============================================================

describe('Register boundaries', () => {
    it('should compile program using r0 and r15', async () => {
        const source = [
            'move r0 0',
            'move r15 0',
            'add r0 r0 r15',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// 数字类型测试
// ============================================================

describe('Number types', () => {
    it('should compile program with all number types', async () => {
        const source = [
            'move r0 42',
            'move r1 3.14',
            'move r2 $FF',
            'move r3 %1010',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// Unicode 标识符测试
// ============================================================

describe('Unicode identifiers', () => {
    it('should compile program with Chinese identifiers', async () => {
        const source = [
            'alias 计数器 r0',
            'alias 设备 d0',
            '主循环:',
            'move r0 0',
            'add r0 r0 1',
            'blt r0 10 主循环',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });
});

// ============================================================
// 大型程序测试
// ============================================================

describe('Large programs', () => {
    it('should compile program with 100+ statements', async () => {
        const lines = ['alias counter r0', 'define LIMIT 100', 'start:', 'move r0 0'];
        for (let i = 0; i < 50; i++) {
            lines.push('add r0 r0 1');
            lines.push('yield');
        }
        lines.push('blt r0 LIMIT start');
        lines.push('hcf');

        const source = lines.join('\n') + '\n';
        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
        expect(result.program.statements.length).toBeGreaterThanOrEqual(100);
    });

    });

// ============================================================
// 常量引用测试
// ============================================================

describe('Constants in instructions', () => {
    // NOTE: 常量关键字(pi/tau/nan等)作为操作数时，program.statements 的 toJSON 序列化
    // 会产生无效 JSON，导致访问 program.statements 抛出 SyntaxError。
    // TODO: 修复 Node 绑定中 AST 序列化常量关键字时产生无效 JSON 的问题
    // 此处通过验证 parser.diagnostics 来确认解析过程正常。
    it('should compile program using pi and tau', async () => {
        const source = [
            'move r0 pi',
            'move r1 tau',
            'mul r0 r0 r1',
            'hcf',
        ].join('\n') + '\n';

        const result = await compile(source);

        expect(result.parser.diagnostics).toHaveLength(0);
    });

        });

// ============================================================
// 注释处理测试
// ============================================================

describe('Comments in programs', () => {
    });

// ============================================================
// 文档注释与类型提示测试
// ============================================================

describe('Doc comments and type hints', () => {
                });

// ============================================================
// 类型推导与语义分析系统测试
// ============================================================

describe('Type inference and semantic analysis', () => {
                        });

File: code/IC10/backend/compiler/tests/java/src/test/java/io/github/stationeers/ic10/SystemTest.java
// Copyright (c) 2026. All rights reserved.
// This source code is licensed under the CC BY-NC-SA
// (Creative Commons Attribution-NonCommercial-NoDerivatives) License, By Xiao Songtao.

package io.github.stationeers.ic10;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

import static org.junit.jupiter.api.Assertions.*;

/**
 * System tests for the IC10 Java binding.
 * <p>
 * Mirrors the C++ {@code test_system.cpp} and Node.js {@code system.test.ts}
 * test suites at matching granularity. Validates end-to-end compilation of
 * real IC10 programs covering realistic scenarios: counting loops, airlock
 * control, gas filtering, complete instruction arities, all preprocessor
 * directives, error scenarios, boundary cases, forward references, HASH/STR
 * macros, device/register boundaries, number types, Unicode identifiers,
 * large programs, constants, comments, doc comments, type hints, and type
 * inference / semantic analysis.
 *
 * @author edocsitahw
 * @since 1.1.0
 */
@DisplayName("System tests")
class SystemTest {

    @BeforeAll
    static void setupLanguage() {
        IC10CompilerLocal.setLanguage("zh-hans");
    }

    // ---- Helpers ----

    /** 编译结果载体：携带各阶段产物 / Compile result carrying pipeline outputs. */
    private static final class CompileResult {
        final Token[] tokens;
        final Program program;
        final Parser parser;
        final Analyser analyser;

        CompileResult(Token[] tokens, Program program, Parser parser, Analyser analyser) {
            this.tokens = tokens;
            this.program = program;
            this.parser = parser;
            this.analyser = analyser;
        }
    }

    /** 完整编译流水线：source → tokens → AST → 语义分析 / Full pipeline. */
    private static CompileResult compile(String source) {
        Token[] tokens = Lexer.tokenize(source, false);
        Parser parser = new Parser();
        parser.init(tokens, false);
        Program program = parser.parse();
        Analyser analyser = new Analyser();
        analyser.visit(program);
        return new CompileResult(tokens, program, parser, analyser);
    }

    /** 读取 grammarTest.ic fixture 文件 / Read grammarTest.ic fixture. */
    private static String readGrammarTestFile() throws IOException {
        String path = System.getProperty("grammarTest.path");
        assertNotNull(path, "grammarTest.path system property must be set");
        return new String(Files.readAllBytes(Paths.get(path)), StandardCharsets.UTF_8);
    }

    /** 判断诊断列表中是否包含指定 ID / Check if diagnostics contain the given ID. */
    private static boolean hasDiagnostic(Diagnostic[] diags, String id) {
        for (Diagnostic d : diags) {
            if (id.equals(d.id)) {
                return true;
            }
        }
        return false;
    }

    // ============================================================
    // 真实 IC10 程序测试
    // ============================================================

    @Nested
    @DisplayName("Real IC10 programs")
    class RealPrograms {

        @Test
        @DisplayName("compile counting loop program")
        void compileCountingLoopProgram() {
            String src = String.join("\n",
                    "alias counter r0",
                    "define LIMIT 10",
                    "start:",
                    "move r0 0",
                    "loop:",
                    "add r0 r0 1",
                    "blt r0 LIMIT loop",
                    "yield",
                    "j start"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics");
            assertNotNull(result.program.toJSON());
        }

        

        
    }

    // ============================================================
    // grammarTest.ic fixture 测试
    // ============================================================

    @Nested
    @DisplayName("grammarTest.ic fixture")
    class GrammarTestFixture {

        @Test
        @DisplayName("compile grammarTest.ic successfully")
        void compileGrammarTestSuccessfully() throws IOException {
            String src = readGrammarTestFile();
            assertFalse(src.isEmpty(), "grammarTest.ic should not be empty");
            CompileResult result = compile(src);
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }

        

        
    }

    // ============================================================
    // 指令元数完整覆盖测试
    // ============================================================

    @Nested
    @DisplayName("All instruction arities")
    class AllInstructionArities {

        @Test
        @DisplayName("compile program with all instruction arities")
        void compileAllArities() {
            // NOTE: lbn/lbns 最后操作数需为 identifier|number，不能用 register
            String src = String.join("\n",
                    "# 零元",
                    "hcf",
                    "yield",
                    "# 一元",
                    "sleep 100",
                    "j main",
                    "# 二元",
                    "move r0 0",
                    "abs r0 r1",
                    "# 三元",
                    "add r0 r1 r2",
                    "sub r0 r1 r2",
                    "mul r0 r1 r2",
                    "# 四元",
                    "lerp r0 r1 r2 r3",
                    "# 五元",
                    "lbn r0 100 200 2 Average",
                    "# 六元",
                    "lbns r0 100 200 3 Quantity Average",
                    "main:",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for all arities");
        }
    }

    // ============================================================
    // 预处理指令完整覆盖测试
    // ============================================================

    @Nested
    @DisplayName("All preprocessor directives")
    class AllPreprocessorDirectives {

        @Test
        @DisplayName("compile program with all preprocessor directive types")
        void compileAllDirectives() {
            String src = String.join("\n",
                    "alias devA d0",
                    "alias regA r0",
                    "define COUNT 42",
                    "define HEX_VAL $FF",
                    "define HASH_VAL HASH(\"StructureLiquidVolumePump\")",
                    "define STR_VAL STR(\"SomeHash\")",
                    "main:",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for all directives");
        }
    }

    // ============================================================
    // 错误场景测试
    // ============================================================

    @Nested
    @DisplayName("Error scenarios")
    class ErrorScenarios {

        @Test
        @DisplayName("report diagnostics for undefined variable")
        void reportUndefinedVariable() {
            CompileResult result = compile("move r0 undefined_var\nhcf\n");
            assertTrue(result.analyser.getDiagnostics().length > 0,
                    "analyser should report undefined variable");
        }

        @Test
        @DisplayName("report diagnostics for redefined alias")
        void reportRedefinedAlias() {
            String src = String.join("\n",
                    "alias foo r0",
                    "alias foo r1",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertTrue(result.analyser.getDiagnostics().length > 0,
                    "analyser should report redefined alias");
        }

        @Test
        @DisplayName("report diagnostics for redefined define")
        void reportRedefinedDefine() {
            String src = String.join("\n",
                    "define MAX 10",
                    "define MAX 20",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertTrue(result.analyser.getDiagnostics().length > 0,
                    "analyser should report redefined define");
        }

        @Test
        @DisplayName("report diagnostics for redefined label")
        void reportRedefinedLabel() {
            String src = String.join("\n",
                    "start:",
                    "move r0 0",
                    "start:",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertTrue(result.analyser.getDiagnostics().length > 0,
                    "analyser should report redefined label");
        }

        @Test
        @DisplayName("report parser diagnostics for missing newline")
        void reportMissingNewline() {
            CompileResult result = compile("hcf yield\n");
            assertTrue(result.parser.getDiagnostics().length > 0,
                    "parser should report missing newline between statements");
        }
    }

    // ============================================================
    // 边界情况测试
    // ============================================================

    @Nested
    @DisplayName("Boundary cases")
    class BoundaryCases {

        @Test
        @DisplayName("handle empty source")
        void handleEmptySource() {
            CompileResult result = compile("");
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }

        @Test
        @DisplayName("handle source with only comments")
        void handleOnlyComments() {
            String src = "# comment 1\n# comment 2\n# comment 3\n";
            CompileResult result = compile(src);
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }

        @Test
        @DisplayName("handle source with only newlines")
        void handleOnlyNewlines() {
            CompileResult result = compile("\n\n\n\n");
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }

        @Test
        @DisplayName("handle source with trailing newlines")
        void handleTrailingNewlines() {
            CompileResult result = compile("hcf\n\n\n\n");
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }

        @Test
        @DisplayName("handle single statement without trailing newline")
        void handleSingleStatementNoTrailingNewline() {
            CompileResult result = compile("hcf");
            assertNotNull(result.program);
            assertNotNull(result.program.toJSON());
        }
    }

    // ============================================================
    // 前向引用测试
    // ============================================================

    @Nested
    @DisplayName("Forward references")
    class ForwardReferences {

        @Test
        @DisplayName("resolve forward label reference in jump instruction")
        void resolveForwardLabelInJump() {
            String src = String.join("\n",
                    "j target",
                    "move r0 1",
                    "target:",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            // 前向引用应被解析，不应有 IEA3 系列未定义错误
            for (Diagnostic d : result.analyser.getDiagnostics()) {
                assertFalse(d.id != null && d.id.contains("IEA3"),
                        "forward reference should be resolved, but got: " + d.id);
            }
        }

        

        @Test
        @DisplayName("resolve multiple forward references")
        void resolveMultipleForwardReferences() {
            String src = String.join("\n",
                    "j label1",
                    "j label2",
                    "j label3",
                    "label1:",
                    "hcf",
                    "label2:",
                    "yield",
                    "label3:",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            for (Diagnostic d : result.analyser.getDiagnostics()) {
                assertFalse(d.id != null && d.id.contains("IEA3"),
                        "forward references should be resolved, but got: " + d.id);
            }
        }
    }

    // ============================================================
    // HASH 和 STR 宏测试
    // ============================================================

    @Nested
    @DisplayName("HASH and STR macros")
    class HashAndStrMacros {

        @Test
        @DisplayName("compile HASH macro in define")
        void compileHashMacro() {
            CompileResult result = compile(
                    "define PUMP_HASH HASH(\"StructureLiquidVolumePump\")\nhcf\n");
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for HASH macro");
        }

        @Test
        @DisplayName("compile STR macro in define")
        void compileStrMacro() {
            CompileResult result = compile(
                    "define LABEL STR(\"SomeHash\")\nhcf\n");
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for STR macro");
        }

        @Test
        @DisplayName("compile multiple HASH and STR macros")
        void compileMultipleMacros() {
            String src = String.join("\n",
                    "define HASH1 HASH(\"StructureLiquidVolumePump\")",
                    "define HASH2 HASH(\"StructureGasVentSensor\")",
                    "define STR1 STR(\"Label1\")",
                    "define STR2 STR(\"Label2\")",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for multiple macros");
        }
    }

    // ============================================================
    // 设备引用测试
    // ============================================================

    @Nested
    @DisplayName("Device references")
    class DeviceReferences {

        @Test
        @DisplayName("compile program using all device references d0-d5")
        void compileAllDeviceReferences() {
            String src = String.join("\n",
                    "l r0 d0 Setting",
                    "l r1 d1 Setting",
                    "l r2 d2 Setting",
                    "l r3 d3 Setting",
                    "l r4 d4 Setting",
                    "l r5 d5 Setting",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for d0-d5");
        }

        
    }

    // ============================================================
    // 寄存器边界测试
    // ============================================================

    @Nested
    @DisplayName("Register boundaries")
    class RegisterBoundaries {

        @Test
        @DisplayName("compile program using r0 and r15")
        void compileR0AndR15() {
            String src = String.join("\n",
                    "move r0 0",
                    "move r15 0",
                    "add r0 r0 r15",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for r0 and r15");
        }
    }

    // ============================================================
    // 数字类型测试
    // ============================================================

    @Nested
    @DisplayName("Number types")
    class NumberTypes {

        @Test
        @DisplayName("compile program with all number types")
        void compileAllNumberTypes() {
            String src = String.join("\n",
                    "move r0 42",
                    "move r1 3.14",
                    "move r2 $FF",
                    "move r3 %1010",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for all number types");
        }
    }

    // ============================================================
    // Unicode 标识符测试
    // ============================================================

    @Nested
    @DisplayName("Unicode identifiers")
    class UnicodeIdentifiers {

        @Test
        @DisplayName("compile program with Chinese identifiers")
        void compileChineseIdentifiers() {
            String src = String.join("\n",
                    "alias 计数器 r0",
                    "alias 设备 d0",
                    "主循环:",
                    "move r0 0",
                    "add r0 r0 1",
                    "blt r0 10 主循环",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for Chinese identifiers");
        }
    }

    // ============================================================
    // 大型程序测试
    // ============================================================

    @Nested
    @DisplayName("Large programs")
    class LargePrograms {

        @Test
        @DisplayName("compile program with 100+ statements")
        void compileLargeProgram() {
            java.util.List<String> lines = new java.util.ArrayList<>();
            lines.add("alias counter r0");
            lines.add("define LIMIT 100");
            lines.add("start:");
            lines.add("move r0 0");
            for (int i = 0; i < 50; i++) {
                lines.add("add r0 r0 1");
                lines.add("yield");
            }
            lines.add("blt r0 LIMIT start");
            lines.add("hcf");
            String src = String.join("\n", lines) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for large program");
            assertNotNull(result.program.toJSON());
        }

        
    }

    // ============================================================
    // 常量引用测试
    // ============================================================

    @Nested
    @DisplayName("Constants in instructions")
    class ConstantsInInstructions {

        @Test
        @DisplayName("compile program using pi and tau")
        void compilePiAndTau() {
            String src = String.join("\n",
                    "move r0 pi",
                    "move r1 tau",
                    "mul r0 r0 r1",
                    "hcf"
            ) + "\n";
            CompileResult result = compile(src);
            assertEquals(0, result.parser.getDiagnostics().length,
                    "parser should produce no diagnostics for pi/tau");
        }

        

        
    }

    // ============================================================
    // 注释处理测试
    // ============================================================

    

    // ============================================================
    // 文档注释与类型提示测试
    // ============================================================

    

    // ============================================================
    // 类型推导与语义分析系统测试
    // ============================================================

    
}
